lib/saml/url_service.rb

```diff
@@ -17,27 +17,33 @@ class URLService
     LOGIN_REDIRECT_PARTIAL = '/auth/login/callback'
     LOGOUT_REDIRECT_PARTIAL = '/logout/'
 
-    attr_reader :saml_settings, :session, :authn_context
+    attr_reader :saml_settings, :session, :authn_context, :query_params
 
-    def initialize(saml_settings, session: nil, user: nil)
+    def initialize(saml_settings, session: nil, user: nil, query_params: {})
       if session.present?
         @session = session
         @authn_context = user&.authn_context
       end
+
       @saml_settings = saml_settings
+      @query_params = query_params
+
+      if query_params[:RelayState]
+        @query_params[:RelayState] = relay_state_params.merge(query_params[:RelayState]).to_json
```

Comment:
yes, i would think we want to be merging into the hash, as RelayState will eventually ALWAYS include at minimum a session_activity_id and request_id.

Reply:
At first this made sense to me, but it seems we'd always be passing it when it matters, as in all of the session#new types will include a `type` param in the RelayState. Always including it will expose the RelayState in the callback URLs, which would serve no purpose.